working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+- [Wikipedia - Structured Concurrency](https://en.wikipedia.org/wiki/Structured_concurrency)
+
+### `dart:concurrent`
+
+`dart:concurrent` will serve as a library hosting low-level concurrency
+primitives.
+
+#### `Thread` and `ThreadPool`
+
+Isolates are _not_ threads even though they are often confused with ones. A code
+running within an isolate might be executing on a dedicated OS thread or it
+might running on a dedicated pool of threads. When expanding Dart's
+multithreading capabilities it seems reasonable to introduce more explicit ways
+to control threads.
+
+```dart
+// dart:concurrent
+
+abstract class Thread implements Shareable {
+  /// Runs the given function in a new thread.
+  ///
+  /// The function is run in a shared isolate, meaning that
+  /// it will not have access to the non-shared state.
+  ///
+  /// The function will be run in a `Zone` which uses the
+  /// spawned thread as an executor for all callbacks: this
+  /// means the thread will remain alive as long as there is
+  /// a callback referencing it.
+  external static Thread start<T>(FutureOr<T> Function() main);
+
+  /// Returns the current thread on which the execution occurs.
+  ///
+  /// Note: Dart code is only guaranteed to be pinned to a specific OS thread during a synchronous execution.
+  external static Thread get current;
+
+  Future<void> join();
+
+  void interrupt();
+
+  external set priority(ThreadPriority value);
+  ThreadPriority get priority;
+}
+
+/// An [Executor] backed by a fixed size thread
+/// pool and owned by the shared isolate of the
+/// current isolate (see [Isolate.runShared]).
+abstract class ThreadPool implements Executor {
+  external factory ThreadPool({required int concurrency});
+}
+```
+
+Additionally I think providing a way to synchronously execute code in a specific
+isolate on the _current_ thread might be useful:
+
+```dart
+class Isolate {
+    T runSync<T>(T Function() cb);
+}
+```
+
+##### Asynchronous code and threads
+
+Consider the following example:
+
+```dart
+Thread.start(() async {
+  var v = await foo();
+  var u = await bar();
+});
+```
+
+Connecting execution / scheduling behavior to `Zone` allows us to give a clear
+semantics to this code: this code will run on a specific (newly spawned) thread
+and will not change threads between suspending and resumptions.
+
+#### Atomic operations
+
+`AtomicRef<T>` is a wrapper around a value of type `T` which can be updated
+atomically. It can only be used with true reference types - an attempt to create
+an `AtomicRef<int>`, `AtomicRef<double>` , `AtomicRef<(T1, ..., Tn)>` will
+throw.
+
+```dart
+// dart:concurrent
+
+class AtomicRef<T extends Shareable> implements Shareable {
+  /// Atomically updates the current value to [desired].
+  ///
+  /// The store has release memory order semantics.
+  void store(T desired);
+
+  /// Atomically reads the current value.
+  ///
+  /// The load has acquire memory order semantics.
+  T load();
```

Comment:
Just make it `abstract T value;`.

Reply:
My motivation for using `load` and `store` instead of getter and setter was to align with extension methods (which use atomicLoad and atomicStore naming) and to indicate that this does something more complicated then just storing a value (e.g. there is a barrier to ensure release semantics).

Reply:
Added a comment explaining this to the doc.